sky/backends/backend_utils.py

```diff
@@ -133,13 +137,69 @@ def fill_template(template_name: str,
         output_path = str(output_path)
     output_path = os.path.abspath(output_path)
 
-    # Add yaml file path to the template variables.
+    # Runtime files handling
+    #
+    # List of runtime files to be uploaded to cluster:
+    #   - yaml config (for autostopping)
+    #   - wheel
+    #   - credentials
+    # Format is {dst: src}.
+    file_mounts = {SKY_RAY_YAML_REMOTE_PATH: output_path}
+
+    # fill_template() is also called to fill TPU/spot controller templates,
+    # which don't have all variables.
+    if 'sky_remote_path' in variables and 'sky_local_path' in variables:
+        file_mounts[variables['sky_remote_path']] = variables['sky_local_path']
+    if 'credentials' in variables:
+        file_mounts.update(variables['credentials'])
+        variables['credentials'] = {}
+    # Putting these in file_mounts hurts provisioning speed, as each file
+    # opens/closes an SSH connection.  Instead, we:
+    #  - cp locally them into a directory
+    #  - upload that directory as a file mount (1 connection)
+    #  - use a remote command to move all runtime files to their right places.
+
+    # yaml config
     variables['sky_ray_yaml_remote_path'] = SKY_RAY_YAML_REMOTE_PATH
     variables['sky_ray_yaml_local_path'] = output_path
+
+    # Local tmp dir holding runtime files.
+    local_runtime_files_dir = tempfile.mkdtemp()
+    variables['local_runtime_files_dir'] = local_runtime_files_dir
+    # Remote dir.
+    variables['remote_runtime_files_dir'] = _REMOTE_RUNTIME_FILES_DIR
+
+    # Build a command that moves runtime files to their right destinations.
+    commands = []
+    for dst, src in file_mounts.items():
+        # Our runtime files (wheel, yaml, credentials) do not have backslashes.
+        assert not src.endswith('/'), src
+        assert not dst.endswith('/'), dst
+        src_basename = os.path.basename(src)
+        dst_basename = os.path.basename(dst)
```

Comment:
Instead of using the basename, we may need to consider renaming the files in the `local_runtime_files_dir` to prevent any potential conflicts. Or we can record all the `src_basename` in a set and make sure there is no conflicts.
```
basenames = set()
for dst, src in file_mounts.items():
  ...
  assert src_basename not in basenames, f'Duplicated src basename: {src_basename}'
  basenames.add(src_basenames)
```

Reply:
Good catch, done.